ecommerce/views.py

```diff
@@ -152,3 +165,47 @@ def post(self, request, *args, **kwargs):
                     )
         # The response does not matter to CyberSource
         return Response()
+
+
+class CouponsView(ListModelMixin, GenericViewSet):
+    """
+    View for coupons API. This is a read-only API showing the user
+    - what coupons they have available if those coupons would be automatically applied on checkout
+    - what coupons they have for a given coupon code, even if those coupons wouldn't be automatically applied
+    """
+    authentication_classes = (
+        SessionAuthentication,
+        TokenAuthentication,
+    )
+    permission_classes = (IsAuthenticated,)
+
+    def get_queryset(self):
+        """List coupons which a user is allowed to see"""
+        return pick_coupons(self.request.user)
+
+    serializer_class = CouponSerializer
+
+
+class UserCouponsView(APIView):
+    """
+    View for coupon/user attachments. Used to create attachments for a user for a coupon.
+    """
+    permission_classes = (IsLoggedInUser, IsAuthenticated,)
+    authentication_classes = (
+        SessionAuthentication,
+        TokenAuthentication,
+    )
+
+    def post(self, request, code, *args, **kwargs):
+        """Attach a coupon to a user"""
+        with transaction.atomic():
+            coupon = get_object_or_404(Coupon, coupon_code=code)
+            if not is_coupon_redeemable(coupon, self.request.user):
+                raise Http404
```

Comment:
I was thinking a 404 would be appropriate since it's basically saying, coupon not found. And the distinction between coupon not found and invalid or unredeemable coupon is irrelevant to the user